src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,617 @@
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method, for example the initial key material. However, if
+ * {@code getProviderName} is called before calling the {@code deriveKey} or
+ * {@code deriveData} methods, the first provider supporting the KDF algorithm
+ * is chosen which may not be the desired one; therefore it is recommended not
+ * to call {@code getProviderName} until after a key derivation operation. Once
+ * a provider is selected, it cannot be changed.
+ * <p>
+ * The {@code deriveKey} and {@code deriveData} methods of {@code KDF} objects
+ * must be thread-safe. That is, multiple threads may concurrently invoke these
+ * methods on a single {@code KDF} object with no ill effects.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDFWithHmacSHA256");
+ *
+ *    KDFParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 42);
+ *
+ *    kdfHkdf.deriveKey("AES", kdfParameterSpec);
+ *}
+ *
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param kdfParameters
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                KDFParameters kdfParameters) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        // the lock is not needed, because the Spi will already be set in
+        // chooseProvider
+        lock = null;
```

Comment:
I guess that by `chooseProvider` you mean `chooseFirstProvider`. However, I'm not sure how not having a lock would work in cases such as this one:

```
KDF kdf = KDF.getInstance("HkdfSHA512", sunPKCS11);
SecretKey derivedKey = kdf.deriveKey("AES", kdfParameterSpec);
```

I'm getting a `NullPointerException` because `lock` is `null` in `KDF::deriveKey`.